esphome/components/voice_kit/__init__.py

```diff
@@ -1,31 +1,125 @@
+import hashlib
+from pathlib import Path
 import esphome.codegen as cg
 import esphome.config_validation as cv
 from esphome import pins
+from esphome import automation, core, external_files
 from esphome.components import i2c
-from esphome.const import CONF_ID, CONF_RESET_PIN
+from esphome.const import (
+    CONF_ID,
+    CONF_RAW_DATA_ID,
+    CONF_RESET_PIN,
+    CONF_URL,
+    CONF_VERSION,
+)
+from esphome.core import HexInt
 
 CODEOWNERS = ["@kbx81"]
 DEPENDENCIES = ["i2c"]
 
+CONF_FIRMWARE = "firmware"
+CONF_MD5 = "md5"
+
+DOMAIN = "voice_kit"
+
 voice_kit_ns = cg.esphome_ns.namespace("voice_kit")
 VoiceKit = voice_kit_ns.class_("VoiceKit", cg.Component, i2c.I2CDevice)
+VoiceKitFlashAction = voice_kit_ns.class_("VoiceKitFlashAction", automation.Action)
+
+
+def _compute_local_file_path(url: str) -> Path:
+    h = hashlib.new("sha256")
+    h.update(url.encode())
+    key = h.hexdigest()[:8]
+    base_dir = external_files.compute_local_file_dir(DOMAIN)
+    return base_dir / key
+
+
+def download_firmware(config):
+    url = config[CONF_URL]
+    path = _compute_local_file_path(url)
+    external_files.download_content(url, path)
+
+    try:
+        with open(path, "r+b") as f:
+            firmware_bin = f.read()
+    except Exception as e:
+        raise core.EsphomeError(f"Could not open firmware file {path}: {e}")
+
+    firmware_bin_md5 = hashlib.md5(firmware_bin).hexdigest()
+    if firmware_bin_md5 != config[CONF_MD5]:
+        raise cv.Invalid(f"{CONF_MD5} is not consistent with file contents")
+
+    return config
+
 
 CONFIG_SCHEMA = (
     cv.Schema(
         {
             cv.GenerateID(): cv.declare_id(VoiceKit),
             cv.Required(CONF_RESET_PIN): pins.gpio_output_pin_schema,
+            cv.GenerateID(CONF_RAW_DATA_ID): cv.declare_id(cg.uint8),
+            cv.Optional(CONF_FIRMWARE): cv.All(
+                {
+                    cv.Required(CONF_URL): cv.url,
+                    cv.Required(CONF_VERSION): cv.version_number,
+                    cv.Required(CONF_MD5): cv.All(cv.string, cv.Length(min=32, max=32)),
```

Comment:
Can we automatically extract these from the files?

Reply:
I know nothing about the XMOS binary format so, if that information is there at all, I don't know how we would extract it.

Reply:
Also, we are using the MD5 value to verify the file before we write it into ESPHome -- it doesn't go onto the device. The XMOS platform appears to have its own integrity verification.